plugin.py

```diff
@@ -105,42 +99,40 @@ def is_server_downloaded():
     return os.path.exists(get_server_exec())
 
 
-class LspElixirPlugin(LanguageHandler):
-    @property
-    def name(self):
-        return __package__.lower()
+class LspElixirPlugin(AbstractPlugin):
+    @classmethod
+    def name(cls):
+        return "elixir"
 
-    @property
-    def config(self):
+    @classmethod
+    def configuration(cls):
         config = {
             "enabled": True,
             "command": [get_server_exec()],
         }
+        config.update(DEAFULT_SETTINGS)
 
-        filename = '{}.sublime-settings'.format(__package__)
-        settings = sublime.load_settings(filename)
-        for key, default in DEAFULT_SETTINGS.items():
-            config[key] = settings.get(key, default)
-
-        return read_client_config(self.name, config)
-
-    def on_start(self, window):
-        if not is_elixir_installed():
-            sublime.status_message(
-                "Please install Elixir for LSP-elixir plugin to work.")
-            return False
+        user_configs, file_path = super().configuration()
+        for key, value in config.items():
+            if not user_configs.has(key):
+                user_configs.set(key, value)
 
-        if not is_server_downloaded():
-            sublime.status_message(
-                "Elixir Language Server not yet downloaded.")
-            return False
+        return user_configs, file_path
 
-        return True
+    @classmethod
+    def needs_update_or_installation(cls):
+        return not is_server_downloaded()
 
+    @classmethod
+    def install_or_update(cls):
+        download_server()
 
-def plugin_loaded():
-    if not is_server_downloaded():
-        download_server_async()
+    def on_pre_server_command(self, command):
+        if command['command'].startswith('editor.action'):
+            return Promise(
```

Comment:
If public API relies on `Promise` then the `Promise` should be exported from public API. @rwols 

But ideally, we would create some abstraction for it so that we can change the underlying implementation from Promise to asyncio in the future without breaking compatibility.

Reply:
But in this case, since the return value is Optional, you should be able to just return `None` and not import `Promise` at all, no?